### PR TITLE
29242 Diagram check followups

### PR DESCRIPTION
**Pull request by instagibbs**
Follow-ups to https://github.com/bitcoin/bitcoin/pull/29242

### CODE REVIEW

File: src/txmempool.cpp [hunk 1]
```diff
@@ -1302,7 +1302,7 @@ util::Result<std::pair<std::vector<FeeFrac>, std::vector<FeeFrac>>> CTxMemPool::
     // Step 1: build the old diagram.
 
     // The above clusters are all trivially linearized;
-    // they have a strict topology of 1 or two connected transactions.
+    // they have a strict topology of one or two connected transactions.
```
Comment (glozow) [inline src/txmempool.cpp:1305]: 836841e2683fa0ca3fd7837726d935e52be142bc seems a bit trivial :sweat_smile:

File: src/test/rbf_tests.cpp [hunk 2]
```diff
@@ -396,6 +396,14 @@ BOOST_FIXTURE_TEST_CASE(improves_feerate, TestChain100Setup)
     // With one more satoshi it does
     BOOST_CHECK(ImprovesFeerateDiagram(pool, {entry1}, {entry1, entry2}, tx1_fee + tx2_fee + 1, tx1_size + tx2_size) == std::nullopt);
 
+    // With prioritisation of in-mempool conflicts, it effects the results of the comparison using the same args as just above
```
Comment (glozow) [inline src/test/rbf_tests.cpp:399]: b380a1df36a95c85310949aeb81d8c57f4d023c4: s/effects/affects

File: src/test/rbf_tests.cpp [hunk 3]
```diff
@@ -37,6 +37,47 @@ static inline CTransactionRef make_tx(const std::vector<CTransactionRef>& inputs
...
+    witness.stack.emplace_back(10);
+    tx1.vin[0].scriptWitness = witness;
+
+    for (size_t i = 0; i < output_values.size(); ++i) {
+        tx1.vout[i].scriptPubKey = CScript() << OP_11 << OP_EQUAL;
+        tx1.vout[i].nValue = output_values[i];
+    }
+
+    // Second tx takes second parent output
+    CMutableTransaction tx2 = CMutableTransaction();
```
Comment (glozow) [inline src/test/rbf_tests.cpp:64]: Could just make a copy of tx1 and change the input prevout? Can remove all the duplicate code.

File: src/test/rbf_tests.cpp [hunk 4]
```diff
@@ -37,6 +37,47 @@ static inline CTransactionRef make_tx(const std::vector<CTransactionRef>& inputs
     return MakeTransactionRef(tx);
 }
 
+// Make two sibling transactions
```
Comment (glozow) [inline src/test/rbf_tests.cpp:40]:
```suggestion
// Make two child transactions from parent (which must have at least 2 outputs).
// Each tx will have the same outputs, using the amounts specified in output_values.
```

File: src/test/rbf_tests.cpp [hunk 5]
```diff
@@ -67,6 +108,21 @@ static CTransactionRef add_descendant_to_parents(const std::vector<CTransactionR
...
+    EXCLUSIVE_LOCKS_REQUIRED(::cs_main, pool.cs)
+{
+    AssertLockHeld(::cs_main);
+    AssertLockHeld(pool.cs);
+    TestMemPoolEntryHelper entry;
+    // Assumes this isn't already spent in mempool
+    auto children_tx = make_two_siblings(/*parent=*/parent, /*output_values=*/{50 * CENT});
+    pool.addUnchecked(entry.FromTx(children_tx.first));
+    pool.addUnchecked(entry.FromTx(children_tx.second));
+    // Return last created tx
```
Comment (glozow) [inline src/test/rbf_tests.cpp:122]: Returns both, actually?

File: src/test/fuzz/rbf.cpp [hunk 6]
```diff
@@ -178,5 +178,16 @@ FUZZ_TARGET(package_rbf, .init = initialize_package_rbf)
...
-    if (!calc_results.has_value()) assert(err_tuple.value().first == DiagramCheckError::UNCALCULABLE);
+    if (!calc_results.has_value()) {
+         assert(err_tuple.value().first == DiagramCheckError::UNCALCULABLE);
+    } else {
+        // Diagram check succeeded
+        if (!err_tuple.has_value()) {
+            // New diagram's final fee should always match or exceed old diagram's
+            assert(calc_results->first.back().fee <= calc_results->second.back().fee);
+        } else if (calc_results->first.back().fee > calc_results->second.back().fee) {
+            // Or it failed, and if old diagram had higher fees, it should be a straight
```
Comment (glozow) [inline src/test/fuzz/rbf.cpp:189]: 4f910a69448203fc51abd993bef27bf00768a1ec "a straight"?

**Review by glozow — COMMENTED** [review]
LGTM, thanks for taking my test nits. I haven't run the package_rbf fuzzer with changes yet. Will review again after rebase.

File: src/test/rbf_tests.cpp (hunk 2 above)
Comment (instagibbs) [reply src/test/rbf_tests.cpp:399]: RFC: merge the two words into "ffect"

File: src/test/fuzz/rbf.cpp (hunk 6 above)
Comment (instagibbs) [reply src/test/fuzz/rbf.cpp]: .... failure, fixed

File: src/test/rbf_tests.cpp (hunk 5 above)
Comment (instagibbs) [reply src/test/rbf_tests.cpp]: copy/paste, removed

File: src/test/rbf_tests.cpp (hunk 3 above)
Comment (instagibbs) [reply src/test/rbf_tests.cpp]: done

File: src/test/rbf_tests.cpp (hunk 2 above)
Comment (instagibbs) [reply src/test/rbf_tests.cpp]: done

File: src/txmempool.cpp (hunk 1 above)
Comment (instagibbs) [reply src/txmempool.cpp]: removed :)

**Comment by instagibbs** [comment]
rebased on master, ready for more review

File: src/test/fuzz/rbf.cpp (hunk 6 above)
Comment (glozow) [reply src/test/fuzz/rbf.cpp]: image

**Comment by glozow** [comment]
```
src/txmempool.cpp:1298:    // all_conflicts.  every such transaction will either be at its own feerate (followed 
src/txmempool.cpp:1299:    // by any descendant at its own feerate), or as a single chunk at the descendant's 

^^^
Trailing whitespace
```

**Comment by instagibbs** [comment]
fixed, thanks

**Review by glozow — COMMENTED** [review]
ACK ee1b9b231a0a7e89b77cbf8ea54e0534f0970dd0, reviewed the changes and package_rbf fuzzer seems to run fine

**Comment by instagibbs** [comment]
@sdaftuar @ismaelsadeeq @willcl-ark :pray:

**Review by murchandamus — COMMENTED** [review]
crACK ee1b9b231a0a7e89b77cbf8ea54e0534f0970dd0

File: src/test/rbf_tests.cpp [hunk 7]
```diff
@@ -67,6 +98,20 @@ static CTransactionRef add_descendant_to_parents(const std::vector<CTransactionR
     return child_tx;
 }
 
+// Makes two children for a single parent
+static std::pair<CTransactionRef, CTransactionRef> add_children_to_parent(const CTransactionRef parent, CTxMemPool& pool)
```
Comment (ismaelsadeeq) [inline src/test/rbf_tests.cpp:102]: nit:
```suggestion
static std::pair<CTransactionRef, CTransactionRef> add_two_children_to_parent(const CTransactionRef parent, CTxMemPool& pool)
```

File: src/test/rbf_tests.cpp [hunk 8]
```diff
@@ -116,6 +161,10 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
     const auto tx12 = make_tx(/*inputs=*/ {m_coinbase_txns[8]}, /*output_values=*/ {995 * CENT});
     pool.addUnchecked(entry.Fee(normal_fee).FromTx(tx12));
 
+    // Will make two children of this single parent
+    const auto tx13 = make_tx(/*inputs=*/ {m_coinbase_txns[9]}, /*output_values=*/ {995 * CENT, 995 * CENT});
```
Comment (ismaelsadeeq) [inline src/test/rbf_tests.cpp:165]: style nit
```suggestion
    const auto tx13 = make_tx(/*inputs=*/{m_coinbase_txns[9]}, /*output_values=*/{995 * CENT, 995 * CENT});
```

File: src/test/rbf_tests.cpp [hunk 9]
```diff
@@ -292,6 +342,14 @@ BOOST_FIXTURE_TEST_CASE(rbf_helper_functions, TestChain100Setup)
...
     BOOST_CHECK_EQUAL(pool.CheckConflictTopology({entry_two_parent_child}).value(), strprintf("%s has 2 ancestors, max 1 allowed", entry_two_parent_child->GetSharedTx()->GetHash().ToString()));
+
+    // Single parent with two children, we will conflict with the siblings directly only
+    const auto two_siblings = add_children_to_parent(tx13, pool);
+    const auto entry_sibling_1 = pool.GetIter(two_siblings.first->GetHash()).value();
+    const auto entry_sibling_2 = pool.GetIter(two_siblings.second->GetHash()).value();
+    BOOST_CHECK_EQUAL(pool.CheckConflictTopology({entry_sibling_1}).value(), strprintf("%s is not the only child of parent %s", entry_sibling_1->GetSharedTx()->GetHash().ToString(), entry13_unchained->GetSharedTx()->GetHash().ToString()));
+    BOOST_CHECK_EQUAL(pool.CheckConflictTopology({entry_sibling_2}).value(), strprintf("%s is not the only child of parent %s", entry_sibling_2->GetSharedTx()->GetHash().ToString(), entry13_unchained->GetSharedTx()->GetHash().ToString()));
+
```
Comment (ismaelsadeeq) [inline src/test/rbf_tests.cpp:352]: style nit: trailing line
```suggestion
```

**Review by ismaelsadeeq — APPROVED** [review]
Code review ACK ee1b9b231a0a7e89b77cbf8ea54e0534f0970dd0

I've tested this locally but havent run the fuzz test.

Feel free to ignore the review comments below, mostly style nits.

**Review by willcl-ark — APPROVED** [review]
ACK ee1b9b231a0a7e89b77cbf8ea54e0534f0970dd0

Amendments look good to me.

**Comment by instagibbs** [comment]
@ismaelsadeeq I'll consider these if I have to touch again

**Pull request merged**